RFC: RadiCLI (#698)
radicle-dev/radicle-link: docs/rfc/0000-cli-infrastructure.adoc

= RFC: RadiCLI
:author: @FintanH
:revdate: 2021-06-04
:revremark: draft
:toc:
:toc-placement: preamble

* Author: {author}
* Date: {revdate}
* Status: {revremark}
* Community discussion: n/a
* Tracking Issue: n/a

== Motivation

RFC-682 set the stage for defining the architecture of applications
built on top of `<<radicle-link>>`. This RFC wishes to build on top of
this specification and fill in the gaps around the CLI component and
its organisation. As dogfooding our own solutions is on our long-term
roadmap, this team wants to provide tooling that they will maintain
and use themselves.

== Overview

> This proposal will outline an binary called `rad`, what it will

Reviewer: ```suggestion
This proposal will outline a binary called `rad`, what it will
```

--------
-- J.R.R. Tolkien, The Return of the King

To allow the ecosystem of commands to be extended by ourselves and
others in the community it will be necessary to extend the `rad`
command with further subcommands. The details of this are outlined in
RFC-682#CLI, however, we will give a quick overview here for the sake
of the reader.

Taking inspiration from tools such as `git` and `cargo`, we propose
that the `rad` binary can be extended by placing another
binary on the users `$PATH` (or `%PATH%` on Windows). This
binary MUST follow the naming convention of `rad-*`. To reiterate
here, these new commands MUST NOT override the core set outlined here
(nor in further RFCs).

== Global Parameters

Many subcommands will likely want to share a set of options. These
options, dubbed as "global parameters" here, will be a defined set --
exported from `radicle-link` for re-use -- that the base `rad` command
will always receive and pass onto the subcommand being called.

Thus, all subcommands SHALL receive these global parameters as part of
their interface. We list these parameters here, but it is NOT an
exhaustive list:

* `verbose`
* `quiet`

> * `profile := active (default) | <profile-id>`

Reviewer: How does this relate to location of the monorepo (i.e. `RAD_HOME`)? Surfacing the relationship would be interesting in the context of this RFC.

Author: Ya, I was thinking about this. I think there's no need to surface `RAD_HOME` since:
a) it can be set on the command-line, i.e. `RAD_HOME=/home/happy/foo rad key create`
b) the general case should be that `XDG_DATA_HOME` is used, and `RAD_HOME` is used in special cases. Contrasting that with the profiles, which can also be set via `RAD_PROFILE`, these will potentially be created and set more frequently

== Online & Offline

In RFC-682#Process-Orchestration, it was decided that the peer-to-peer
damon would not be a long-running process. The terminology we will use
here is that the daemon is *online* if it is running and connected to
the network. Otherwise, if it is not running then it is *offline*.

This disctinction becomes important when thinking about the
functionality of `rad` subcommands. The majority of subcommands that
interact with the storage can be done while the daemon is offline. The
only action the author can currently think of — that is online — is

> replication of a URN from the network.

Reviewer: Is that due to how the current "waiting room" is implemented as an online API call to issue a "request" for an URN? In the spirit of the p2p node converged over the state of the storage is there a potential way forward where requesting/replicating a URN from the network could be done through pure manipulation of the storage in contrast to rely on a running node process?

Author: Well, fundamentally the point is that you need to actually have the daemon running so that it can replicate from another peer (waiting room or not).

As pointed out in the discussion https://github.com/radicle-dev/radicle-link/pull/698/files#r649924840 there are potentially two steps to the process. We do an offline track of a new URN, and when we're online it finds providers and replicates if it finds any.

So I guess the question is: Are there any online commands or is it all offline and the daemon knows what to do when it comes online via the "queded tasks" that we made while offline?

Reviewer 2: Perhaps. But you may still want to `start` / `stop` / `Status` it =>

Author: > But you may still want to start / stop / Status it =>

Right, I was holding off on any specification for daemon control commands, but I think this has convinced me that we want to define them here :) 

It also feels like it should be a requirement that all commands work offline and the daemon can perform the tasks when it's online. The following diagram must commute:
![image](https://user-images.githubusercontent.com/3503528/122535008-087c7280-d01b-11eb-99b4-93c40b8b6d0c.png)
(image created using [quiver](https://q.uiver.app/))

This leaves us in a predicament of thinking about how multiple Radicle
applictions may interact with each other. For example, if Upstream is
running and I create a project via the CLI, then a good user
experience is that the Upstream is updated with the new project
(without refreshing). In RFC-682, it is alluded that a PubSub
component could solve this, however, that requires the daemon to be

> online. The author prefers not to specify the PubSub component

Reviewer: Could be I miss an angle here, but the PubSub as described in 682 is dependent on an "online" p2p node. In contrast it specifically recommends a substrate which would be a system service. Or is your meaning here that the node needs to be running to observe the message from the PubSub?

Author: > In contrast it specifically recommends a substrate which would be a system service.

I'm not sure what you mean here. Could you maybe rephrase it?

> Or is your meaning here that the node needs to be running to observe the message from the PubSub?

Ya, exactly. In the example, if I created a project via the CLI and I also have Upstream running (without the daemon running), then there's no way for Upstream to know that an event happened (the project being created).

Reviewer 2: If upstream is running, it should connect to the pubsubbub, onto which the CLI places a message after creating the project. No daemon involved

Author: I thought we hadn't landed on whether pubsubflub was a standalone service or if it was only up along with the daemon? If it's always available then I guess this section is pointless :upside_down_face: But it might be useful to have a section on events being emitted, what do yis think?

Reviewer 2: I think we weren't sure what the right tool is (D-Bus being rather baroque). But I'm fairly sure that we want some sort of standalone thing, because it can become confusing quickly to see what actually happened. For example:

Say you unplugged your WiFi cable, so whenever the daemon comes up, it won't be able to connect to the network. Now say you have Upstream running, but you do something using another program (eg. push some changes using `git`). What should happen?

One option is that the daemon keeps running (because Upstream is connected to it), gets notified by `git` that something happened, and then sends `updated-refs ...` to Upstream, followed by `not-connected`. The other option is that the daemon exits when it can't connect (a few times). It will miss the notification from `git`, and Upstream will just wait in vain for some kind of event which tells it that something was replicated to the network.

Both seem pretty terrible (as are most "event driven" systems) -- you never know if and when the next message arrives, and "no message" can carry information. We know, however, that we need to record the modification of the repo _either way_, such that the daemon can pick it up once it gains connectivity.

The missing piece here -- and this is not the first time I'm saying this -- is that we can't have "offline first" and "it behaves like GitHub" at the same time. In other words, the act of synchronising the local state with the network must be controllable by the user (and is genuinely an "online" operation, btw). 

So in the example: 

If Upstream is running, receives a notification that the repo state has changed, and decides to pretend "it behaves like GitHub", it is _Upstream's_ responsibility to call a `sync` RPC on the daemon, and synchronously wait for the result. That `sync` RPC obviously needs to be idempotent, and it should probably be possible to retrieve information about when and what the last sync did.

If Upstream is _not_ running, things are slightly more complicated, because we can't have `git push` to also atomically `sync` (the refs must have been written already, and there's no atomic rollback). We can do this non-atomically, though, and just funnel back the result to `git` through the sideband. A `rad git push` wrapper would allow to make it optional (by either passing `--sync` or not).

So in summary, there are zero assumptions one can have about the availability of other processes in the service mesh. Therefore, events are always ephemeral, and it is up to the receiver to act upon them, not the sender. Actions which return a result must be synchronous RPCs. 

A consequence of this is that there must be ways to synchronously inspect state / history in case events were missed. The tricky thing about this is to beware of side-effects: say you want to get some info about the most recent `sync`. If you activate the daemon for that, it will have the side-effect of also attempting to connect to the network -- which may be pointless, because of the WiFi cable situation. In that sense, journaling event logs to disk as @alexjg  suggested might be worth considering.

Reviewer 3: Yeah this is increasingly looking to me like we need a log and not a pubsub system. With a log we don't need to make any assumptions about what systems are running when producing or consuming events, we don't need to abstract over any platform capabilities, and we can more easily implement transactional semantics where we need to.

The downside is that consumers become more complex, but not enormously so I think, we don't need to build Kafka. Consumers just need to store an  offset and handle log file rotations when reading from the log files.

Reviewer 2: That is phrased _slightly_ on the optimistic side 😅

Reviewer 3: Ha, I suppose it is. I'm not suggesting that it's a trivial endeavour to build a log, just that a) it's not the same challenge as building a distributed log and b) I think that architecturally it concentrates the complexity in one place, rather than having the complexity of handling ephemeral messages and disappaering services spread around the various components of the system.

Reviewer 2: Yeah I’m with you. My concern is just that reliable message buses typically lead to YOLOKAFKA syndrome, meaning that everyone is relying on this component to never lose any data.

I’d favour to first try and _reduce_ the amount of bookkeeping required. For example, once #653 lands, it would not strictly be necessary to remember what went out previously. You can just force a sync with your bootstrap nodes, and let git figure out if any work needs to be performed.

Perhaps we should start with pubsub = `ff01::1`?

Author: Thanks for the elaboration and discussion! I think I misinterpreted a previous conversation and was under the assumption the PubSub component was inherently connected to a running daemon, but as you've pointed out that won't be the case.

So in terms of this RFC, it seems like the assumption that will be made is that there will be a PubSub component. If a CLI subcommand deems an event as interesting to track it SHALL publish this event to the PubSub, following the guidelines of the future PubSub RFC. With a caveat being that "interesting" doesn't fall into YOLOKAFKA syndrome -- I'm not entirely sure how we can prevent that but I suppose since it's the consumer's responsibility to sync then they are also responsible for ignoring message. Does that make sense?

`online` the command will see if the daemon is online, and if not ask
it to come online before applying the command.

== Key Access

RFC-682#Signatures outlined the signing of Radicle data. This is essential to
this RFC as commands that write to the storage generally need access
to the secret key, whether for signing data or proof of ownership to
the storage. For most (if not all) read-only commands, no access to
the secret key is needed.

To achieve the goals of having a read-only storage and using
`<<ssh-agent>>` for signing, we propose the following should be
implemented:

. `ssh-agent` compatability with `<<radicle-keystore>>`, which has been
started in `<<rk-17, radicle-keystore#17>>`.
. A subcommand `rad-key-ssh-add` for adding one's secret key to the
`ssh-agent` using the supplied constraints, e.g. ask for password.
. `radicle-link#461` to have a read-only storage.

== Library/Command Organisation

As proposed here, `radicle-link` will provide a core set of
functionality. Something this document wishes to standardise is the
organisation of these components. The author sees two options:
. One binary that defines the core subcommands

> . Many binary that take advantage of the subcommand mechanism

Reviewer: Dunno if GH tripping up or if the formatting generally doesn't produce a proper list here.

Author: Ya, I need a new line to activate the bulleting :)

> . Many binary that take advantage of the subcommand mechanism

Reviewer: ```suggestion
. Many binaries that take advantage of the sub-command mechanism
```

Author: Thanks for the spot :) I'm going to stick with "subcommand" and "subcommands" because I think it's pretty well recognised

Pros of *1.*:

* One-stop shop for all core functionality
* One binary is packaged and distributed

Cons of *1.*:

* Making changes to one subset of the commands could mean a version
  change of the whole binary

Pros of *2.*:

* We will be organising commands as libraries (as per RFC-682) — so
  modular commands would mirror this structure well.
* Subcommands can be versioned independently of each other.
* We can dogfood our subcommand functionality

Cons of *2.*:

* Makes it easier to override one of the core commands.
* Many binary must be packaged and distributed.

The author recommends *1.*. As mentioned, we have a hard requirement
of not allowing the core commands to be overrided. In *2.* this is
_always_ possible.

=== `radicle-bins`

Per RFC-682#CLI, all subcommands will be provided as
libraries. However, this leaves the question of where the binary might
live. So far, it has been an unspoken formalism that all binaries live
in `radicle-bins` where we can provide a `Cargo.lock` file. All
functionality will live in `radicle-link`, including the `main`
function, and the entry in `radicle-bins` will be a thin wrapper

> around this function.

Reviewer: It would be nice to scrutinise this more and challenge if this played out well for the seed.

`radicle-link` and `<<radicle-bins>>` as well as providing the first
iteration of the core set of subcommands.

== One `rad` to Rule Them All

[quote]
-------
One ring to rule them all, one ring to find them,
One ring to bring them all, and in the darkness bind them;
-------
-- J.R.R. Tolkein, Fellowship of the Ring

The integral part of this proposal is that there will be one binary,
`rad`, that will be the foundation for all other subcommands. It will
define the global parameters and how to extend it.

We will discuss these in the following two sections.

== The Fellowship of the `rad`

[quoute]
--------
“I can’t carry it for you, but I can carry you.” – Sam Gamgee
--------
-- J.R.R. Tolkien, The Return of the King

To allow the ecosystem of commands to be extended by ourselves and
others in the community it will be necessary to extend the `rad`
command with further subcommands. The details of this are outlined in
RFC-682#CLI, however, we will give a quick overview here for the sake
of the reader.

Taking inspiration from tools such as `git` and `cargo`, we propose
that the `rad` binary can be extended by placing another

> binary on the users `$PATH` (or the Windows equivalent). This

Reviewer 2: `%PATH%`

binary MUST follow the naming convention of `rad-*`. To reiterate
here, these new commands MUST NOT override the core set outlined here
(nor in further RFCs).

== Global Parameters

Many subcommands will likely want to share a set of options. These
options, dubbed as "global parameters" here, will be a defined set --
exported from `radicle-link` for re-use -- that the base `rad` command
will always receive and pass onto the subcommand being called.

> Thus, all subcommands SHALL receive these global parameters as part of

Reviewer 2: It would be good to detail _how exactly_ they receive those parameters. Builtin commands are obviously just function calls, but how is this handled for `execve`'d binaries? Does the `rad` wrapper parse (and validate) those options? Does it then just output them verbatim to the subcommand? Does it prefix them with `--global-` or something? Is order significant? Or does it use some other format like `CBOR` or `JSON`?

Author: I (hopefully) addressed this in https://github.com/radicle-dev/radicle-link/pull/698/commits/8a3d264dfbd3fd44cef3b27f29c3d7b07536a73c

Reviewer 2: So if I’m not careful, I might accidentally expect a parameter which is already global 🤔 

Follow up question: is `println` the only contract for program output?

Author: > So if I’m not careful, I might accidentally expect a parameter that is already global

RTFM :grimacing: 

Ya, I was thinking what's the point of a global parameter if someone is just going to ignore it and use their own meaning of it. But given some more thought, it would be a Bad Thing:tm: if their behaviour was overridden by the globals. However, if `rad` prefixed them then it would need to pass both parameters, i.e. `--global-profile=abc` and `--profile=abc`, just in case there was a non-global version of the parameter defined. One would hope that wouldn't mess up a parser :crossed_fingers:

> Follow up question: is println the only contract for program output?

It hadn't crossed my mind. The other contract of the output would be the format. Is there anything else that I should think about here?

Reviewer 2: > The other contract of the output would be the format.

Yeah that's the thing. It did cross my mind that @cloudhead (_just_ as an example) might fancy some ASCII spinners, emoji admonitions or whatever. Now, where would you override those? And how do you adhere to the style in an external subcommand?

Maybe this kind of thing is just not feasibly possible.

Author: Right, I think we should only be committing to plumbing commands here for now. I would say external commands can add plumbing on top of that and it's up to them how they mangle `stdout` :smile:

Reviewer 4: Hmm, what if globals were passed as env vars to the sub-command, after being parsed?

Reviewer 2: May, maybe it would still be good to define some kind of machine-readable output format at some point, which can be turned on with `—emoji-bufs` or something.

exhaustive list:

* `verbose`
* `quiet`
* `profile := active (default) | <profile-id>`

== Online & Offline

In RFC-682#Process-Orchestration, it was decided that the peer-to-peer
damon would not be a long-running process. The terminology we will use
here is that the daemon is *online* if it is running and connected to
the network. Otherwise, if it is not running then it is *offline*.

This disctinction becomes important when thinking about the
functionality of `rad` subcommands. The majority of subcommands that
interact with the storage can be done while the daemon is offline. The
only action the author can currently think of — that is online — is
replication of a URN from the network.

This leaves us in a predicament of thinking about how multiple Radicle
applictions may interact with each other. For example, if Upstream is
running and I create a project via the CLI, then a good user
experience is that the Upstream is updated with the new project
(without refreshing). In RFC-682, it is alluded that a PubSub
component could solve this, however, that requires the daemon to be
online. The author prefers not to specify the PubSub component

> here. To bridge the gap, we recommend that a subcommand can add a

Reviewer 2: This is not necessary, and arguably not something the user can or should know. The daemon will start _automatically_ as soon as a command requires it, via socket activation.

started in `<<rk-17, radicle-keystore#17>>`.
. A subcommand `rad-key-ssh-add` for adding one's secret key to the
`ssh-agent` using the supplied constraints, e.g. ask for password.
. `radicle-link#461` to have a read-only storage.

== Library/Command Organisation

As proposed here, `radicle-link` will provide a core set of
functionality. Something this document wishes to standardise is the
organisation of these components. The author sees two options:
. One binary that defines the core subcommands
. Many binary that take advantage of the subcommand mechanism

Pros of *1.*:

* One-stop shop for all core functionality
* One binary is packages and distributed

Cons of *1.*:

* Making changes to one subset of the commands could mean a version
  change of the whole binary

Pros of *2.*:

* We will be organising commands as libraries (as per RFC-682) — so
  modular commands would mirror this structure well.
* Subcommands can be versioned independently of each other.
* We can dogfood our subcommand functionality

Cons of *2.*:

* Makes it easier to override one of the core commands.
* Many binary must be packaged and distributed.

> The author recommends *2.*. The modularity provides a clean separation

Author: The author was on the fence, so happy to recommend 2 =]

of concerns and avoids a monolothic approach. We assume good faith
of overrides not happening [TODO: is there some way we can just

> disallow this programatically?]. For packaging and distribution of the

Reviewer 2: Instead, you may detail how development can be made more convenient. Eg. make it so development builds always produce binaries with a `-dev` suffix, and provide `systemd`/`launchd` definitions so they either call `cargo build` or reference the `-dev` binaries.

Author: Ah so, there would be the `rad` binary, but then there would also be a `rad-key-dev` binary, for example?

Reviewer 2: Thinking about it, maybe the RFC is not the right place to talk about this (the same thing could also be achieved using `nix` and / or `direnv`). Maybe just assume proper packaging.

binaries, we propose to have a script that will install each
subcommand binary — until we have a more well-founded way of
distributing Radicle artifacts.

=== `radicle-bins`

Per RFC-682#CLI, all subcommands will be provided as
libraries. However, this leaves the question of where the binary might
live. So far, it has been an unspoken formalism that all binaries live
in `radicle-bins` where we can provide a `Cargo.lock` file. All
functionality will live in `radicle-link`, including the `main`
function, and the entry in `radicle-bins` will be a thin wrapper
around this function.

=== Distribution

> We are noting here that -- at the time of writing -- we cannot publish

Reviewer 2: Is this relevant? We are relying on the system package manager either way.

these binaries to <<crates.io>> due to the fact that `librad` and other
crucial libraries only live as a `git` repositories and are not on
crates.io. Anything that depends on them cannot be published as a
consequence.

This means that all subcommands will need to be built from source
initially, unless we build the binaries and distribute them from a
well-known source.

== Core

With all the details in place, we define a non-exhaustive list of the
core commands that we wish to include for the first iteration. More
components MAY be added as the result of future RFCs. We SHALL keep a
list of the core commands as part of the `radicle-link` documentation

> — as part of the `README`.

Reviewer 2: ```suggestion

```

> === `rad-project`

Reviewer 2: I don't think project/person as separate subcommands is a smart idea, because that will just repeat most of the commonalities. It also prevents extensibility without having to change the CLI for the extension specifically. 

A more maintainable approach would be to either disambiguate with a `--type` argument, or add another layer of sub-command-ness (eg. `rad-identities project [create|...]`). Or else, move CRUD-style verbs to the top-level, like `rad create project [opts]`.

Author: > It also prevents extensibility without having to change the CLI for the extension specifically.

Can you elaborate on what you mean here?

The rest makes sense, but I'm still mulling over what the consequences of one option over the other might be :) I think I'll need to sketch them out to see which makes more sense

Reviewer 2: I mean that extension payloads are basically opaque JSON blobs. One may also want to pipe an entire payload as JSON, and have the CLI infer the type. Or maybe some new identity kind was introduced, but the local CLI installation doesn’t know about it yet (disregard for a moment that the replication logic needs to know about it — maybe that namespace was obtained by some other means). We could still make it so the CLI is able to display _something_.

Author: Ah that's clear, thanks for the insight!

This command will provide management of the `Project` identities and
their respective working copies. The following actions will be
defined:

* `create` — create a `Project` identity and either create a
  respective working copy, or link it to an existing working copy.

> * `get` — attempt to retrieve a `Project` identity from the store based on

Reviewer 2: What is the output encoding of this and similar commands? Should the user be able to choose (eg. `--output=[json,cbor,java.io.Serializable]`)?

Author: Good point, I had that written down in one of my notes from way back, so thanks for reminding me ^_^ I think it should be like your example says: `--output=` and have a series of options. You forgot `dhall` :P

Reviewer 2: `sexp`

  a provided URN.
* `list` — list all `Project` identities in the store.
* `update` — update a `Project` identity with a new payload.

> * `merge` — merge a peer's version of the `Project` identity with your own

Reviewer 2: This should receive some more thought (apart from `merge` not yet footgun-free): it is possibly preferable to force the user to _review_ what is being accepted, unless a `--force` flag or something is given.

> * `track` — track a peer for the `Project` identity and add the peer

Author: I'm going back and forth on having "tracking" be part of `Project`/`Person` and having a separate component that does this. 

I'm also questioning the working copy being only associated with `Project` after the brief discussion:

> fintohaps:
> Had a thought today: if a `Person` identity was "checked out" as a working copy you could store something like a static webpage in `heads` and then that could be something like your personal blog :)

> kim:
> The difference between Person and Project is only in the delegation rules, and whether some metadata makes sense or not. Otherwise, a namespace is a git repository, and you can put anything you want in there.

> * `track` — track a peer for the `Project` identity and add the peer

Author: Some other things to be added are:
* Listing peers
* Listing delegates
* Get `rad/self`

Author: There might be more that I'm forgetting =]

to the working copy's include file.

=== `rad-person`

This command will provide the management of the `Person`
identities. The following actions will be defined:

* `create`— create a `Person` identity.
* `get`— attempt to retrieve a `Person` identity from the store based
  on a provided URN.
* `list`— list all `Person` identities in the store.
* `update`— update a `Person` identity with a new payload.
* `merge`— merge a peer's version of the `Person` identity with your own.
* `track`— track a peer for the `Person` identity.

=== `rad-replicate`

> * `ask` — ask the network for a given URN and replicate it to local storage.

Reviewer 2: Those are two things.

Reviewer: Would it be possible to track a URN and have the p2p stack take care of replicating it from the network?

Author: > Would it be possible to track a URN and have the p2p stack take care of replicating it from the network?

There's an issue for that actually =] https://github.com/radicle-dev/radicle-link/issues/141

Reviewer 2: Contributions welcome. Essentially, we could just treat a remote without a `url` as a clone intent. Requires some work to decouple from libgit2

  a provided URN.
* `list` — list all `Project` identities in the store.
* `update` — update a `Project` identity with a new payload.
* `merge` — merge a peer's version of the `Project` identity with your own
* `track` — track a peer for the `Project` identity and add the peer
to the working copy's include file.

=== `rad-person`

This command will provide the management of the `Person`
identities. The following actions will be defined:

* `create`— create a `Person` identity.
* `get`— attempt to retrieve a `Person` identity from the store based
  on a provided URN.
* `list`— list all `Person` identities in the store.
* `update`— update a `Person` identity with a new payload.
* `merge`— merge a peer's version of the `Person` identity with your own.
* `track`— track a peer for the `Person` identity.

=== `rad-replicate`

* `ask` — ask the network for a given URN and replicate it to local storage.

=== `rad-key`

This command will provide the management of keys and the `Profile`
construct in `librad`. The following actions will be defined:

* `create`— generate a new key and a new `Profile`, which implies a
  separate storage.

> * `profile get`— get the current `Profile`'s identifier

Reviewer 4: Are these sub-sub-commands? Or is this like, `rad profile-get`? I wonder if for plumbing commands we should keep it simple with only one level of sub-commands, like git, eg. `git branch -D` to delete a branch, not `git branch delete`

Author: What's the advantage of one over the other?

Reviewer 4: I guess the git-style namespacing has these advantages:

* It's clear what's a sub-command vs just a parameter to a command, so it may lead to less ambiguities, eg. `git push origin master` is `git-push origin master` and not `git-push-origin master` or `git-push-origin-master`, especially for commands that accept positional arguments.
* You don't need to handle cases like `rad profile --flag get` vs. `rad profile get --flag`

So basically it's simpler from an implementation perspective to only have 1 level of sub-commands. The advantages of having multiple levels are:

* For CLIs with hundreds of sub-commands and a clear hierarchy, eg. `gcloud`, it may be easier to namespace everything
* It "reads" better maybe or has the promise of being more user-friendly, though I'd say that is subjective

Author: I've leaned towards the hierarchy of commands. I feel that subcommands are easier on the cognitive load rather than trying to remember what shorts were given for a command. I was also thinking that tab completion would work better for subcommands, which isn't possible with flags (or at least not that I know of).
